pkg/workflow/controller/handlers/workflowTrigger/cronJob.go

```diff
@@ -0,0 +1,161 @@
+package workflowTrigger
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/caicloud/cyclone/pkg/k8s/clientset"
+	"github.com/pkg/errors"
+	"github.com/robfig/cron"
+	"github.com/satori/go.uuid"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/caicloud/cyclone/pkg/apis/cyclone/v1alpha1"
+)
+
+const (
+	KeyTemplate = "%s/%s"
+)
+
+type CronTrigger struct {
+	Cron                *cron.Cron
+	IsRunning           bool
+	SuccCount           int
+	FailCount           int
+	Namespace           string
+	WorkflowTriggerName string
+	WorkflowRun         *v1alpha1.WorkflowRun
+	Manage              *CronTriggerManager
+}
+
+type CronTriggerManager struct {
+	Client         clientset.Interface
+	CronTriggerMap map[string]*CronTrigger
+	mutex          sync.RWMutex
+}
+
+func NewTriggerManager(client clientset.Interface) *CronTriggerManager {
+	return &CronTriggerManager{
+		Client:         client,
+		CronTriggerMap: make(map[string]*CronTrigger),
+		mutex:          sync.RWMutex{},
+	}
+}
+
+func getKeyFromTrigger(trigger *CronTrigger) (string) {
+	return fmt.Sprintf(KeyTemplate, trigger.Namespace, trigger.WorkflowTriggerName)
+}
+
+func (m *CronTriggerManager) AddTrigger(trigger *CronTrigger) {
+	wftKey := getKeyFromTrigger(trigger)
+	m.mutex.RLock()
+	if wft, ok := m.CronTriggerMap[wftKey]; ok {
+		m.mutex.RUnlock()
+		// this situation may not happen for k8s will do same name check
+		log.Warnf("failed to add cronTrigger, already exists: %+v\n", wft)
+	} else {
+		m.mutex.RUnlock()
+
+		m.mutex.Lock()
```

Comment:
It's error prone to add two locks here. When you already have read lock, can you acquire write log here ? It will result deadlock in my opinion.

Strongly recommend to only use one lock, in your case "m.mutex.Lock()   defer m.mutex.Unlock()"